### PR TITLE
Tagging with Entities

### PR DESCRIPTION
**Pull request by ML-Dirk**

### CODE REVIEW

File: lambda_functions/ml_service/template.yaml [hunk 1]
```diff
@@ -159,6 +159,46 @@ Resources:
...
+      MemorySize: 832
+      Role:
+        Fn::GetAtt:
+          - LambdaExecutionRole
+          - Arn
+      Events:
+        Echo:
+          Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
+          Properties:
+            Path: /factchecks/{item_id}
```
Comment (SandroGlueck) [inline lambda_functions/ml_service/template.yaml:191]: URL path does not seem to be correct.

Something like /tags?item_id=<item_id> seems more fitting.
item_id has then to be read from queryPathParameters obviously.

File: lambda_functions/ml_service/EnrichItem.py [hunk 2]
```diff
@@ -280,6 +281,68 @@ def store_itementities(event, context, is_test=False, session=None):
...
+        except Exception:
+            # store tag in database
+            tag.id = str(uuid4())
+            tag.tag = str_tag
+            try:
+                update_object(tag, is_test, session)
+            except Exception as e:
+                logger.error(
+                    "Could not store tag. Exception: %s", e, exc_info=True)
+                raise
```
Comment (SandroGlueck) [inline lambda_functions/ml_service/EnrichItem.py:326]: Flow of code is error prone in my opinion as the exception might have other reasons than the item not existing.

I'd rather catch the specific exception that occurs if a tag doesn't exist or let the `get_tag_by_content` function return null if a tag doesn't exist and then continue from there on.

File: lambda_functions/ml_service/EnrichItem.py [hunk 3]
```diff
@@ -280,6 +281,68 @@ def store_itementities(event, context, is_test=False, session=None):
...
+        except Exception:
+            itemtag.id = str(uuid4())
+            itemtag.item_id = item_id
+            itemtag.tag_id = tag.id
+            try:
+                update_object(itemtag, is_test, session)
+            except Exception as e:
+                logger.error(
+                    "Could not store item tag. Exception: %s", e, exc_info=True)
+                raise
```
Comment (SandroGlueck) [inline lambda_functions/ml_service/EnrichItem.py:343]: See previous comment

File: lambda_functions/ml_service/template.yaml (hunk 1 above)
Comment (ML-Dirk) [reply lambda_functions/ml_service/template.yaml:191]: Ups ;-)

File: lambda_functions/ml_service/EnrichItem.py (hunk 2 above)
Comment (ML-Dirk) [reply lambda_functions/ml_service/EnrichItem.py]: get_tag_by_content now returns None if the tag is not existing

File: lambda_functions/ml_service/EnrichItem.py (hunk 3 above)
Comment (ML-Dirk) [reply lambda_functions/ml_service/EnrichItem.py]: get_itemtag_by_tag_and_item_id now returns None if the itemtag is not existing

File: lambda_functions/ml_service/template.yaml [hunk 4]
```diff
@@ -68,7 +68,7 @@ Resources:
         Echo:
           Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
           Properties:
-            Path: /factchecks/{item_id}
+            Path: /tags/{item_id}
```
Comment (SandroGlueck) [inline lambda_functions/ml_service/template.yaml:71]: I still dislike the idea of the `item_id` being a pathParameter.

Strictly following the REST standard, the path is usually meant to identify a document or a collection of documents (in this case a tag or multiple tags). https://restfulapi.net/resource-naming/

If we want to follow the REST standard, I would propose one of the following path templates:

- /items/{item_id}/tags
- /tags?item_id={item_id}

The later would make it necessary to also change the GetTags function to read the item_id from the queryStringParameters. (see [create_review.py](https://github.com/DetektivKollektiv/python-lambda-functions/blob/develop/lambda_functions/review_service/create_review.py#L31) for an example)

Comment (SandroGlueck) [reply lambda_functions/ml_service/template.yaml:71]: @PotSchmoell any feelings?

Comment (ML-Dirk) [reply lambda_functions/ml_service/template.yaml:71]: Thanks for the link :-) I give the first option a try

**Review by SandroGlueck — APPROVED** [review]

File: lambda_functions/ml_service/template.yaml (hunk 4 above)
Comment (PotSchmoell) [reply lambda_functions/ml_service/template.yaml]: I think I would prefer query parameters (option 2). 
They are also rather easy to implement. Check "detektivkollektiv-review_api-get_review_question" in the review service for reference.

**Pull request merged**